packages/core-plugin-api/src/adaptable-components/AdaptationProvider.tsx

```diff
@@ -0,0 +1,260 @@
+   */
+  excludePlugins?: BackstagePlugin<any, any>[];
+
+  /**
+   * Exclude all previously registered adaptations for this/these component(s)
+   */
+  excludeComponents?:
+    | AdaptableComponentRef<any, any>
+    | AdaptableComponentRef<any, any>[];
+
+  /**
+   * Exclude these specific previously registered adaptations
+   */
+  exclude?: ComponentAdaptation[];
+};
+
+type WrapAdaptation<T> = T & { adaptation: ComponentAdaptation<any, any> };
+
+type PropsInterceptor<Props extends {}> = WrapAdaptation<{
+  propsInterceptor: AdaptableComponentPropsInterceptor<Props>;
+}>;
+
+type Adaptation<Props extends {}, Context extends {}> = WrapAdaptation<{
+  component: AdaptableComponentAdaptation<Props, Context>;
+}>;
+
+/** The component ref provider key is its id */
+type AdaptationMapKey = string;
+/**
+ * The component ref provider value is a pre-flattened list of props
+ * interceptors and adaptation components
+ * */
+type AdaptationMapValue<Props extends {}, Context extends {}> = {
+  propsInterceptors: PropsInterceptor<Props>[];
+  components: Adaptation<Props, Context>[];
+};
+
+/**
+ * The context stores a map between the component ref and an object with a list
+ * of props-interceptors and providers.
+ *
+ * A map and pre-flattened arrays that makes it very fast to lookup whether a
+ * component has adaptations, and if so, iterating its props-interceptors and
+ * providers.
+ */
+type ContextType = {
+  adaptationMap: Map<AdaptationMapKey, AdaptationMapValue<any, any>>;
+};
+
+const VersionedContext = createVersionedContext<{ 1: ContextType }>(
+  'adaptation-context',
+);
+
+/**
+ * Provides an {@link @backstage/core-plugin-api#ApiHolder} for consumption in
+ * the React tree.
+ *
+ * @public
+ */
+export function AdaptationProvider(
+  props: PropsWithChildren<AdaptationProviderProps>,
+) {
+  const {
+    adaptations,
+    reset = false,
+    excludePlugins,
+    excludeComponents,
+    exclude,
+    children,
+  } = props;
+  const parentContext = useContext(VersionedContext)?.atVersion(1);
+
+  const contextValue = useMemo((): ContextType => {
+    const adaptationMap = reset
+      ? cloneAdaptationMap(undefined)
+      : cloneAdaptationMap(parentContext?.adaptationMap);
+
+    if (excludeComponents) {
+      const excludeComponentRefs = Array.isArray(excludeComponents)
+        ? excludeComponents
+        : [excludeComponents];
+
+      excludeComponentRefs.forEach(ref => {
+        adaptationMap.delete(ref.id);
+      });
+    }
+
+    if (excludePlugins || exclude) {
+      const excludePluginsSet = new Set(excludePlugins ?? []);
+      const excludeSet = new Set(exclude?.map(e => e.key) ?? []);
+
+      [...adaptationMap.values()].forEach(value => {
+        value.components = value.components
+          .filter(
+            ({ adaptation }) =>
+              !adaptation.plugin || excludePluginsSet.has(adaptation.plugin),
+          )
+          .filter(({ adaptation }) => excludeSet.has(adaptation.key));
```

Comment:
Am I misunderstanding here or should this condition be flipped? 
`.filter(({ adaptation }) => !excludeSet.has(adaptation.key));`
same with line 145 above?

Reply:
You're absolutely right, thanks!

All of this needs unit tests for sure, I've waited with that to get the API bounced here for discussion first though.